### DISCUSSION TITLE
Unable to Publish Blazor WASM project with Quest PDF in GitHub Actions pipeline

### DISCUSSION THREAD

**Discussion by fingers10** [QuestPDF/QuestPDF#806] [Q&A] [answered]
I'm trying to making QuestPDF work in blazor wasm. All was working fine in local. I was able to run and check and its working fine.

However when I tried to publish it in GitHub Actions, I got the following error,

```
System.TypeInitializationException: The type initializer for 'QuestPDF.Fluent.Document' threw an exception.
       ---> QuestPDF.Drawing.Exceptions.InitializationException: The QuestPDF library has encountered an issue while loading one of its dependencies. This type of error often occurs when the current runtime is missing necessary NuGet packages. 
      
      Please ensure the following NuGet packages are added to your primary/startup project and has matching versions:
      - SkiaSharp.NativeAssets.Linux.NoDependencies
      - HarfBuzzSharp.NativeAssets.Linux
      
      For a detailed understanding, please examine the inner exception and consult the official SkiaSharp library documentation.
       ---> System.DllNotFoundException: Unable to load shared library 'libSkiaSharp' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
      /home/runner/.nuget/packages/blazorwasmprerendering.build/3.1.0-preview.4/tools/net8.0/.webhost/libSkiaSharp.so: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.2/libSkiaSharp.so: cannot open shared object file: No such file or directory
      /home/runner/.nuget/packages/blazorwasmprerendering.build/3.1.0-preview.4/tools/net8.0/.webhost/liblibSkiaSharp.so: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.2/liblibSkiaSharp.so: cannot open shared object file: No such file or directory
      /home/runner/.nuget/packages/blazorwasmprerendering.build/3.1.0-preview.4/tools/net8.0/.webhost/libSkiaSharp: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.2/libSkiaSharp: cannot open shared object file: No such file or directory
      /home/runner/.nuget/packages/blazorwasmprerendering.build/3.1.0-preview.4/tools/net8.0/.webhost/liblibSkiaSharp: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.2/liblibSkiaSharp: cannot open shared object file: No such file or directory
```

I have added the needed dependencies mentioned in the above error. But still the issue persists. Please can you assist me on what I'm doing wrong? Here is the link for pipeline failure - https://github.com/ILoveDotNet/ilovedotnet/actions/runs/8130785653/job/22219400410

So after trying all the suggested answers it was still not working, I have commented the code and continued as shown below.

```
<!-- <ItemGroup>
  <NativeFileReference Include="$(HarfBuzzSharpStaticLibraryPath)/3.1.34/st/*.a" />
</ItemGroup> -->

<ItemGroup>
<PackageReference Include="BlazorWasmPreRendering.Build" Version="3.1.0-preview.4" />
<!-- <PackageReference Include="HarfBuzzSharp.NativeAssets.Linux" Version="7.3.0.1" /> -->
	  <!-- <PackageReference Include="HarfBuzzSharp.NativeAssets.WebAssembly" Version="7.3.0.1" /> -->
<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="8.0.2" />
<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="8.0.2" PrivateAssets="all" />
<PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
<PackageReference Include="Microsoft.Extensions.Logging.Configuration" Version="8.0.0" />
<PackageReference Include="PublishSPAforGitHubPages.Build" Version="2.1.1" />
<PackageReference Include="Serilog.Extensions.Logging" Version="8.0.0" />
<PackageReference Include="Serilog.Settings.Configuration" Version="8.0.0" />
<PackageReference Include="Serilog.Sinks.BrowserConsole" Version="2.0.0" />
<PackageReference Include="Serilog.Sinks.BrowserHttp" Version="1.0.0-dev-00032" />
<!-- <PackageReference Include="SkiaSharp.HarfBuzz" Version="2.88.7" /> -->
<!-- <PackageReference Include="SkiaSharp.NativeAssets.Linux.NoDependencies" Version="2.88.7" /> -->
	  <!-- <PackageReference Include="SkiaSharp.NativeAssets.WebAssembly" Version="2.88.7" /> -->
      <!-- <PackageReference Include="SkiaSharp.Views.Blazor" Version="2.88.7" /> -->
</ItemGroup>
```

Any inputs would be highly appreciated. here is the repo link - https://github.com/ILoveDotNet/ilovedotnet

Code changes related to QuestPDF is done in `ReportDemoComponents/ReportDemoComponents.csproj` and `Web/Web.csproj`

**Comment by MarcinZiabek** [comment, marked as answer]
Apologies for the long delay — I am currently revisiting older discussions to make sure none are left without an answer 😄 

Looking at the stack trace, the failure did not occur in the browser runtime but during the prerendering step: BlazorWasmPreRendering.Build executes the application on the Linux build host, and that host process could not locate the native libSkiaSharp.so binary within the tool's probing paths. This was a native-asset resolution limitation of the prerendering tool rather than an issue in QuestPDF itself.

That said, the situation has changed significantly since this discussion was created. Starting with the 2024.3 release, QuestPDF no longer depends on SkiaSharp and instead ships its own native rendering layer, so the packages mentioned in the error message are no longer relevant. The library currently targets desktop and server environments (Windows, Linux, macOS), and the browser-wasm runtime is not supported. For Blazor applications, the recommended architecture is to generate documents server-side (e.g., via an API endpoint or Blazor Server) and deliver the file to the browser.

Since the discussion concerns an architecture that is no longer in use, I am closing it as outdated. Please feel free to open a new discussion if you have any questions regarding the current versions of the library.
